File: linked-list/java/singly_linked_list.java/linked_list.java
class LinkedList<T>
{
    Node<T> head;
    int size;

    LinkedList() {
        head = null;
        size = 0;
    }

    void prepend(Node<T> nodeToPrepend)
    {
        nodeToPrepend.nextNode = head;
        
        head = nodeToPrepend;
        
        size++;
    }

    void append(Node<T> nodeToAppend)
    {
        if(isEmpty()) {
            prepend(nodeToAppend);
            return;
        }
        
        Node<T> currentNode = head;
        
        while(currentNode.nextNode != null) {
            currentNode = currentNode.nextNode;
        }

        currentNode.nextNode = nodeToAppend;

        size++;
    }

    void insertAfterNode(Node<T> nodeToInsertAfter, Node<T> nodeToInsert)
    {
        if(isEmpty()) {
            prepend(nodeToInsert);
            return;
        }

        nodeToInsert.nextNode = nodeToInsertAfter.nextNode;
        nodeToInsertAfter.nextNode = nodeToInsert;
        size++;
    }

    void deleteNode(Node<T> nodeToDelete)
    {
        if(isEmpty()) {
            return;
        }

        Node<T> currentNode = head;

        if(currentNode == nodeToDelete) {
            head = head.nextNode;
            return;
        }

        while(currentNode.nextNode != nodeToDelete) {
            currentNode = currentNode.nextNode;
        }
        currentNode.nextNode = nodeToDelete.nextNode;
        size--;
    }

    Node<T> findNode(T value)
    {
        if(isEmpty()) {
            return null;
        }

        Node<T> currentNode = head;
        
        while(currentNode != null) {
            if(currentNode.data == value) {
                return currentNode;
            }
            currentNode = currentNode.nextNode;
        }

        return null;
    }

    boolean isEmpty()
    {
        return head == null;
    }

    void print()
    {
        Node<T> currentNode = head;
        
        System.out.print("HEAD -> ");
        
        while(currentNode != null) {
            System.out.print(currentNode.data + " -> ");
            currentNode = currentNode.nextNode;
        }
        
        System.out.print("NULL\n");
    }

    int getSize() {
        return size;
    }
}

class Main
{
    public static void main(String[] args) {

        LinkedList<Integer> linkedList = new LinkedList<Integer>();
        
        for (int i = 0; i < 5; i++) {
            linkedList.prepend(new Node<Integer>(i+1));
        }

        linkedList.prepend(new Node<Integer>(100));
        linkedList.append(new Node<Integer>(78));
        linkedList.insertAfterNode(linkedList.findNode(3), new Node<Integer>(102));
        linkedList.insertAfterNode(linkedList.findNode(1), new Node<Integer>(11));
        linkedList.insertAfterNode(linkedList.findNode(78), new Node<Integer>(101));
        linkedList.deleteNode(linkedList.findNode(1));
        linkedList.deleteNode(linkedList.findNode(11));
        linkedList.deleteNode(linkedList.findNode(101));

        System.out.println("Size is: " + linkedList.getSize() + "\n");
        linkedList.print();
    }
}

File: linked_list/linked_list.c
#include <stdlib.h>
#include <stdio.h>

struct node {
    int value;
    struct node *nextNode;
};

typedef struct node node_t;

void print_nodes(node_t *head) {
    node_t *temporary = head;
    printf("HEAD -> ");
    while (temporary != NULL) {
        printf("%d -> ", temporary->value);
        temporary = temporary->nextNode;
    }
    printf("NULL\n");
}

int isEmpty(node_t *head) {
    return head == NULL ? 1 : 0;
}

node_t *create_new_node(int value) {
    node_t *node = malloc(sizeof(node_t));
    node->value = value;
    node->nextNode = NULL;
    return node;
}

void insert_node_at_head(node_t **pointerToHead, node_t *node_to_insert) {
    node_to_insert->nextNode = *pointerToHead;
    *pointerToHead = node_to_insert;
}

void insert_node_at_tail(node_t *head, node_t *node_to_insert) {
    if(1 == isEmpty(head)) {
        return;
    }

    node_t *tmp = head;
    while(tmp->nextNode != NULL) {
        tmp = tmp->nextNode;
    }
    tmp->nextNode = node_to_insert;
}

node_t *find_node(node_t *head, int value) {
    if(1 == isEmpty(head)) {
        return NULL;
    }

    node_t *tmp = head;
    while(tmp->nextNode != NULL) {
        if(tmp->value == value) {
            return tmp;
        }
        tmp = tmp->nextNode;
    }
    return NULL;
}

void insert_node_after_node(node_t *node_to_insert_after, node_t *node_to_insert) {
    if(1 == isEmpty(node_to_insert_after)) {
        return;
    }

    node_to_insert->nextNode = node_to_insert_after->nextNode;
    node_to_insert_after->nextNode = node_to_insert;
}

void delete_node(node_t *head, node_t *node_to_delete) {
    if(1 == isEmpty(head)) {
        return;
    }

    node_t *tmp = head;
    while(tmp->nextNode != node_to_delete) {
        tmp = tmp->nextNode;
    }
    tmp->nextNode = node_to_delete->nextNode;
}

int main() {

    node_t *head = NULL;

    for (int i = 0; i < 10; ++i) {
        insert_node_at_head(&head, create_new_node(i+1));
    }
    print_nodes(head);

    insert_node_after_node(find_node(head, 6), create_new_node(15));
    print_nodes(head);

    insert_node_at_tail(head, create_new_node(100));
    print_nodes(head);

    delete_node(head, find_node(head, 2));
    print_nodes(head);

    if(NULL != find_node(head, 6)) {
        printf("Node was found.");
    } else {
        printf("Node not found.");
    }

    return 0;
}